### DISCUSSION TITLE
Memory and updates

### DISCUSSION THREAD

**Discussion by infromthecold** [maziac/DeZog#59] [Ideas]
Hi again
           A few things I have noticed would be nice, if possible.
           
1 When I try and change the variables or program counter manually it has no affect on the live machine, so it would be nice if that was possible. 

2 When debugging with a memory view, again on the live machine it does not update the view as we step through our code, is that possible?

Do these available on cSpect?

**Comment by maziac** [comment]
Hi,
Actually both should be possible.
Not 100% sure about the PC,have to try myself.
But to change the PC you can also set the cursor on a line in your source code and do a right-click. You then have an option to 'set PC to cursor'.

What is not possible: You cannot change variables in the Watches panel.

What is your setup? I.e. what is 'live machine'? A real ZXNext connected via serial IF?

The memory view should reflect all changes that are done on memory. Changes that are done manually in the memory view directly and changes that are done by your z80 code.

Could you explain more what you are doing?

Could you also send me the version numbers of all involved SW.

All of this should also work on cspect and zesarux.

**Reply by infromthecold** [reply]
Wow thanks for a quick response.

Found the run from cursor that will be handy, my set up is a spectrum next with the cable for the joystick port, the memory view does update when starting with the registers -rmv switch, but not when the memory window from the debug console -mv $c000 $400.

When i am doing is trying to trace a linked list in memory where the addresses change as i link and unlink the links, but the memory window does not reflect the changes while stepping, to see the memory updates i have to close the memory window and open it again which makes debugging harder.

I would try it on cspect but the I cant get cspect to open my game files but that is a cspect issue and not a dezog!

**Reply by infromthecold** [reply]
sorry DeZog-1.4.10 i think, on windows 10

**Reply by infromthecold** [reply]
to be clear, when starting with the -rmv switch mv $c000 $400 does show live changes but when starting without the switch mv $c000 $400 does not reflect live changes

**Reply by maziac** [reply]
First: Congratulations. I think not many people have manage to get the complex setup with a real ZXNext to work.

But, are you sure about the DeZog version?
Current version is 2.2.4.

Could you maybe try to check if it is working if you choose a different memory area, e.g. 8000h instead of c000h.
I had to do a lot of memory swapping on the ZXNext, so maybe there is a bug in somewhere for a certain slot.

For cspect, this is the command line I use for example:
mono CSpect.exe -w4 -zxnext  -nextrom -exit -brk -tv -r -v -debug  -mmc=../../your-rootdir/

For window you would need to remove the "mono" and probably might need to change forward slashes to back slashes.
The "-mmc=" lets you directly use the following directory as your root dir. But make sure you shouldn't have any spaces after the "=". And I think the trailing slash is also mandatory.

**Reply by maziac** [reply]
Ah, OK. I think our last comments crossed.
Will try that.

**Reply by maziac** [reply]
Could you send me your launch.json file? (Or maybe even the whole project so that I could try myself).

**Reply by maziac** [reply]
Was able now to reproduce this. But only once in a while.
I have no clear pattern when this occurs.
For me it happened now twice for a "-mv" without any "-rmv" before. But unfortunately most of the time it does work.
At the moment I'm not able to reproduce the behavior.

Would be nice to have a reproducible way to get to this situation.

**Reply by maziac** [reply]
Seems I'm getting closer.
I can reproduce the behavior now. It happens on a real ZXNext and on Cspect, even with the internal Simulator.

It happens only if a debug session has been restarted with the 'green circle arrow'.
It does not happen if you close the debug session (with the red square) and start a new one with the green triangle.

@infromthecold could you please verify if this works for you: 
(Don't use -rmv). Start a debug session. Use "-mv". Step, see if the memory view gets updated. Stop debug session with red square. Start new debug session. Use "-mv".  Step, see if the memory view gets updated. 
If I'm right this should work.

If you (in the above sequence) use the green circled arrow instead of stopping with the red square, the memory view is not updated.

**Reply by infromthecold** [reply]
Sorry i currently have my head in the debugging, so i cant test it just yes as i have the machine in mid debug, I think its happening when you have multiple memory windows open too, i think they are all not getting updated!

I thought it may be how much data but I only have 3 memory windows open each only 4 bytes, 

-mv $cc71 4
-mv $c000 4
-mv $cc4e 4

**Comment by maziac** [comment]
Please try the fix here:
https://github.com/maziac/DeZog/releases/tag/v2.2.5-test
Install via command palette ('Extensions: Install from VSIX...').
And let me know if it helped with the memory views.

**Reply by infromthecold** [reply]
Much better, that is working, even with the 3 windows open as mentioned above!

**Comment by maziac** [comment]
OK, fine that it is working.

What about the problem with the VARIABLES (i.e. registers)?
Does it also exist still?

I tried to change the register in the VARIABLES panel but couldn't find any problem.
I could even change the PC there and it works.

**Reply by infromthecold** [reply]
Is the only way to change a variable is in a memory window? I am used to Visual studio where you can change them in a watch window. 

As for the registers, I have realised i need to double clicking on the register, and clear the string back to the hex value "9B52h" to make it work as changing just the hex value with the rest "9B52h, 39762u, -25774i, animateDelete" does not work, which was what i was trying.

Maybe for convenience you could just display the hex when editing

**Reply by maziac** [reply]
Oh, I see, you tried to change variables in the memory window.
There is an easier way:
Press the '+' in the WATCH window.
Add the name of one of your labels (e.g. mylabel).
If the size is not recognized correctly you can enter it as well.
E.g. "mylabel,2,1" for a word sized variable.
Please see the DeZog Help chapter 8.11 for more info.

For the registers:
If you double click you can change the registers. The complete field gets selected and you can enter a new number. After you changed the number and press enter the register is changed.
Up to the point when you press enter, the complete input is handled by vscode. I.e. I cannot influence the behavior.

**Comment by infromthecold** [comment]
I think there is some confusion, yes i can add and edit the watches, but i
can't alter the values of the watched variable during a debug session like
we can the register.

thanks for yesterday's update, with the new version it made it easier to
debug

**Reply by maziac** [reply]
Yes, that's right. You cannot alter values in the Watches panel.

Unfortunately vscode does not offer an interface for it.

**Reply by maziac** [reply]
You can try also in other languages: In the vscode WATCH panel you cannot alter values of variables.

What I could do is to provide a debug command for it (like the "-mv"). E.g.:
"-set yourlabel newvalue"

**Reply by maziac** [reply]
Could you please install "dezog-2.2.5-1.vsix" from here:
https://github.com/maziac/DeZog/releases/edit/v2.2.5-test

I added a new command "-memset". Please use the help ("-h") to get some examples for the syntax.
You can write bytes or words in little endian or big endian and even fill complete memory areas.

**Comment by infromthecold** [comment]
Can't we hold ctrl down, so we can alter them like the registers, and
behind the scenes you alter the memory?

**Reply by maziac** [reply]
I don't think I have any direct access to the keyboard.

**Comment by infromthecold** [comment]
That's great, thank you, I will get on it in the morning

Thanks

**Reply by infromthecold** [reply]
Yup that works a treat 

thank you

**Comment by infromthecold** [comment]
Hi Maziac
I tried using the move program counter to the cursor today and it appears
not to be working during debug sessions, i think it thinks the cursor is
where the highlighted line for debugging is. However if I manually set the
PC to an address it works!

Cheers

**Reply by maziac** [reply]
?
Could you elaborate this a little bit.
"Move Program Counter to Cursor" is working **only** during a debug session.
I.e. if you are not inside a debug session the menu is not available on right-click.

If you set the cursor to a line with an assembly statement and execute "Move Program Counter to Cursor" the PC is changed to that address.
This does not  work if the cursor is not on a line with an assembly statement. E.g. if it is on an empty line above.

Do you maybe have some screenshots to illustrate the behavior?

**Comment by infromthecold** [comment]
Sorry for the delay getting back to you, it appears to be working how you
described

Many thanks

**Comment by maziac** [comment]
Hi, could you have a look at a new release here (2.2.5-2):
https://github.com/maziac/DeZog/releases/edit/v2.2.5-test

I have renamed the commands, so please use the "-h" help.

Furthermore I added a memory view that displays data in units of words.
In big and little endian.
Use "-mvw".
Let me know if it works for you.

**Comment by infromthecold** [comment]
Yup will tomorrow

**Reply by infromthecold** [reply]
That's great thanks, -msetb does the job, just noticed -eval too thats going to be real handy as that was something i was missing,  the "-e gr" is "Error: Error: not supported" on a real next, is that correct? 

If you are looking for something to do, I was wondering if it was possible for us to call the serial loop in your code so we can get live updates of some watches or simply to let us add a break point to running code. I think I remember there was something like that in PDS development system back in the day.

But so far its going great,  many thanks

**Reply by maziac** [reply]
"-e gr": "gr" is a zesarux specific command. It is not supported in the other Remotes.
"-e|exec" in general is to send Remote specific commands. But since only zesarux supports one for text commands, it is in fact only for zesarux specific commands.

**Reply by maziac** [reply]
Your other suggestion:
1. The ZXNext environment is very special. I have had a lot of headache with the design of the Next's serial interface. Main problems were: it does not support an interrupt, i.e. it needs to be polled, and it is shared with the joysticks.
That there is no support for interrupt is the main reason why you can't break a running program for vscode's UI. You always have to press the key at the ZXNext to do so. There is a whole chapter in Usage.md which explains the caveats.
So, when there is no possibility to communicate with the Next while the program is running, it is also not possible to retrieve information from it.
2. For the other Remote's it is a little bit different because a running program can be interrupted. However, here is a different problem: vscode's UI does not support such a behavior. Whenever the program is running the VARIABLES pane etc. is turned off. This also happens for other debuggers for other languages. I.e. even if the data would be updated it would not be visible.

What can you do instead:
Have you looked at the LOGPOINT feature? It allows you to log output (registers, memory contents) whenever a certain address is executed.
Instead you can also use a vscode log point (see same chapter in Usage.md) on the fly, while you are debugging.
But drawback with these log points is that they are completely handled by DeZog. I.e. behind the scenes it creates a breakpoint. When hit it evaluates the action, then grasps memory and register values from the Remote, evaluates the expression and prints it.
Afterwards it continues execution.
I.e. it slows down the program execution.
But the good thing is that you can use rather advanced expressions to print.

**Comment by infromthecold** [comment]
Ahhh thanks for that detailed response.

**Comment by maziac** [comment]
If you like you can try another update (defog-2.2.5-3.vsix):
https://github.com/maziac/DeZog/releases/edit/v2.2.5-test

I thought about the drawback in the WATCH panel that the values cannot be modified and
came up with a solution.
Modification is not possible in the WATCH panel but it is in the VARIABLEs panel.
So I added a new command "-addvar expression". 'expression' has the same syntax as what you would enter in the WATCH panel.
But the data now appears in the VARIABLE panel and if you double click an item you can modify the value.

![Bildschirmfoto 2021-08-10 um 18 50 26](https://user-images.githubusercontent.com/3081973/128900781-3c347c29-8dfa-47ce-bdb2-f291e3e3cd6e.jpg)

**Comment by infromthecold** [comment]
Hey, sorry about the delay getting back to you, however i can't get this to
work on the real next,

I can add the variable from the debug console using -addvar
(playerLives),2,1, then i find it in labels (which i think will confuse
some), then i double click it to change the value and I get this error,
when I enter a new value, i get the same error when i select "set value"
from the menu

[image: dezog.png]

Also there appears to be no --addlabel or -removelable command making the
error redundant

Error: Unknown command: '-removeLabel playersLives' and Error: Unknown
command: '-removeLabel (playersLives)'
Error: Unknown command: '-addLabel playersLives' and Error: Unknown
command: '-addLabel (playersLives)'

very confusing, also there is no -removevar command either, the watches are
permanent between sessions however the vars/labels are not which would be a
aonoying having to set them all but every debug sessions

**Reply by maziac** [reply]
Hi,

meanwhile I also implemented the complement command "-rmvar index" to remove a label.  You have to use the index here, i.e. 0 will remove the first displayed label and so on.
Get the new dezog-2.2.5-4.vsix from here again: https://github.com/maziac/DeZog/releases/edit/v2.2.5-test
About the naming. I was also considering "-addlabel/rmlabel" maybe it's the better name. What do you think?
(BTW I chose addvar because it is added to the VARiables pane, but I'm also in doubt that this is a good choice.)


About the error: it happens when the element count is 1. This bug is still in. As a workaround for now you can use:
-addvar playerLives,2,2
I.e. choose 2 for the last parameter instead of 1.

**Comment by infromthecold** [comment]
I don't like label, as its variables we will be altering, I know they are
technically labels but I'm old school. 😁

Thanks for the update I will look tomorrow.

Oh also any chance of adding a dissembly at cursor rather than just program
counter, as sometimes I want to see the code somewhere other than the PC.

**Comment by infromthecold** [comment]
Both "-addvar playersLives,2,2" and  "-addvar playersLives,1,2" worked a
treat and they were editable, also the "-rmvar index" did its job too.

I was also able to use "-addvar playersLives+2,2,2" which was nice

Nice work cheers

**Comment by maziac** [comment]
Nice to hear it works.

About the name:
I was also thinking about "addexpr" for adding an expression, as it is not limited to labels only.
For the variables panel I'm currently thinking as well about "Expressions" or many "Globals" and put it beneath the "Local Stack".
Any other opinion on that would be great to hear.
(BTW I don't really like "rmvar" as it looks very much like the other command "rmv").

For the disassembly:
You can also use the "-dasm" command. E.g.:
~~~
-dasm 400h 5
0400	SBC A,B 
0401	RRA 
0402	PUSH AF 
0403	INC B 
0404	AND C 
~~~
Instead of 400h you can also use a label.

A disassembly at cursor position would also be possible.
Would it be OK to output it in the debug console as well?
Problem here is that it is not possible to enter a count of lines, so maybe it would need to be restricted to only 1 line.

**Comment by maziac** [comment]
No hurry, take your time.

If you like here is a new version (2.2.5-5):
https://github.com/maziac/DeZog/releases/edit/v2.2.5-test

The commands are now "-addexpr" and "-delexpr".
In the VARIABLES section I called it "Epressions".
Furthermore there is a right-click command to disassemble at cursor position.
Also the restart (green circle arrow) is now really a restart only. I.e. all the memory views etc. that you opened in the session are not closed.

**Comment by infromthecold** [comment]
In the morning
